sound/soc/sof/sof-audio.c

```diff
@@ -350,13 +349,30 @@ int sof_widget_list_setup(struct snd_sof_dev *sdev, struct snd_sof_pcm *spcm, in
 	/* set up widgets in the list */
 	for_each_dapm_widgets(list, num_widgets, widget) {
 		struct snd_sof_widget *swidget = widget->dobj.private;
+		struct snd_sof_widget *pipe_widget;
 
 		if (!swidget)
 			continue;
 
-		ret = sof_widget_setup(sdev, swidget);
+		/* set up the pipeline widget */
+		pipe_widget = swidget->pipe_widget;
+		if (!pipe_widget) {
+			dev_err(sdev->dev, "error: no pipeline widget found for %s\n",
+				swidget->widget->name);
+			ret = -EINVAL;
+			goto widget_free;
+		}
+
+		ret = sof_widget_setup(sdev, pipe_widget);
```

Comment:
why we have this dependency? can't we treat the pipe_widget as a normal swidget? the pipe_widget is in the list also no?

Reply:
> @keyonjie we dont connect the pipeline widget to anything topology do we? So, it wont show up in the list of connected widgets right?

We do, iirc. it's easy to confirm it with printing out all widgets in the list, can you check that?

Reply:
im not leaving anything out here @keyonjie. The list of connected DAPM widgets does not contain the pipeline widget. It was rhetorical question. I know for a fact that the pipeline widget is standalone in tplg and not connected to anything.

Reply:
> im not leaving anything out here @keyonjie. The list of connected DAPM widgets does not contain the pipeline widget. It was rhetorical question. I know for a fact that the pipeline widget is standalone in tplg and not connected to anything.

yes, you are right, just checked the pipeline widget is in sdev->widget_list but not in the DAPM widget list.